### PR TITLE
Support migration with "-incoming defer"

### PR DESCRIPTION
**Pull request by YongxueHong**
Support migration with "-incoming defer".

ID: 1766663
Signed-off-by: Yongxue Hong yhong@redhat.com

### CODE REVIEW

**Comment by YongxueHong** [comment]
> Hello @YongxueHong, I briefly reviewed it and more importantly tested it and thank you, this is what I was looking for for a long time smile. The only issue that caught my attention is the lack of HMP interface, would you please include something like this?
> 
> ```diff
> diff --git a/virttest/qemu_monitor.py b/virttest/qemu_monitor.py
> index eae8d40d..941c0f65 100644
> --- a/virttest/qemu_monitor.py
> +++ b/virttest/qemu_monitor.py
> @@ -1387,6 +1387,17 @@ class HumanMonitor(Monitor):
>          """
>          return self.cmd("migrate_set_speed %s" % value)
>  
> +    def migrate_incoming(self, uri):
> +        """
> +        Start an incoming migration, the qemu must have been started
> +        with -incoming defer.
> +
> +        :param uri: The Uniform Resource Identifier identifying the
> +                    source or address to listen on.
> +        :type uri: str
> +        """
> +        return self.cmd("migrate_incoming %s" % uri)
> +
>      def migrate_set_downtime(self, value):
>          """
>          Set maximum tolerated downtime (in seconds) for migration.
> ```
> 
> It works for me...

Hi @ldoktor 
Yes, thanks for reminding, I will update it.
Thanks.

**Comment by ldoktor** [comment]
@wainersm @ovoshchana you should be interested in this patch...

File: virttest/qemu_vm.py [hunk 1]
```diff
@@ -4270,6 +4279,13 @@ def migrate(self, timeout=virt_vm.BaseVM.MIGRATE_TIMEOUT, protocol="tcp",
                         raise exceptions.TestError(msg)
 
             logging.info("Migrating to %s", uri)
+            if (protocol in ('rdma', 'x-rdma', 'tcp', 'unix') and
+                    Flags.INCOMING_DEFER in self.devices.caps):
```
Comment (luckyh) [inline virttest/qemu_vm.py:4283]: Looks the following approach is more reasonable.
```python
deferral_incoming = False

    ...
    if Flags.INCOMING_DEFER in self.devices.caps:
        incoming_val = ' -incoming defer'
        deferral_incoming = True
    qemu_command += incoming_val

...

logging.info("Migrating to %s", uri)
if deferral_incoming:
    ...
```

Comment (YongxueHong) [reply virttest/qemu_vm.py]: Thanks, have updated.

File: virttest/qemu_vm.py [hunk 2]
```diff
@@ -1440,6 +1440,9 @@ def sort_key(dev):
         if (params.get("qemu_force_use_drive_expression", "no") == "yes" and
                 Flags.BLOCKDEV in devices.caps):
             devices.caps.clear_flag(Flags.BLOCKDEV)
+        if (params.get("qemu_force_use_non_defer_expression", "no") == "yes" and
```
[the remark this answers is not in the document]
Comment (YongxueHong) [reply virttest/qemu_vm.py]: Looks good to me and updated.

**Comment by YongxueHong** [comment]
Hi, @wangxianxian @leexiaohui32
Please help me to review, thanks.

File: virttest/qemu_vm.py [hunk 3]
```diff
@@ -4270,6 +4282,12 @@ def migrate(self, timeout=virt_vm.BaseVM.MIGRATE_TIMEOUT, protocol="tcp",
                         raise exceptions.TestError(msg)
 
             logging.info("Migrating to %s", uri)
+            if self.deferral_incoming:
```
[the remark this answers is not in the document]
Comment (ldoktor) [reply virttest/qemu_vm.py:4285]: (What I meant to add is that currently the first VM is not started with migrate_type, therefore it does not check the support for defer so currently this check is always False even though the clone uses defer, which is why it currently does not work at all)

[the remark this answers is not in the document]
Comment (YongxueHong) [reply virttest/qemu_vm.py]: Hi @ldoktor 
Thanks for reviewing, I forgot it, here should be corresponding with the following `clone.monitor.migrate_incoming(_uri)`. so should check the `clone` flag.

**Comment by wangxianxian** [comment]
I have tried this patch to run some related cases on ppc platform, the results are all pass both for "qemu_force_use_static_incoming_expression" is yes and no.
cases are as following:
migrate.default.tcp.default
migrate.default.tcp.with_filter_off.with_post_copy
migrate.default.mig_cancel
migrate.default.unix.default
migrate.default.exec.gzip_exec
ACK.

**Comment by YongxueHong** [comment]
Hi, @leexiaohui32
Please help me to review, thanks.

**Comment by leexiaohui32** [comment]
ack after test on x86 with and without "--customsparam=qemu_force_use_static_incoming_expression=yes":
```
[root@hp-dl385g10-14 kar]# python3 ConfigTest.py --testcase=migrate.default.tcp.default,migrate.default.tcp.with_filter_off.with_post_copy,migrate.default.mig_cancel,migrate.default.unix.default,migrate.default.exec.gzip_exec --guestname=RHEL.8.3.0 --clone=no
 (1/5) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.migrate.tcp: PASS (174.82 s)
 (2/5) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.migrate.tcp.with_filter_off.with_post_copy: PASS (62.87 s)
 (3/5) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.migrate.unix: PASS (150.09 s)
 (4/5) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.migrate.exec.gzip_exec: PASS (199.88 s)
 (5/5) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.migrate.mig_cancel: PASS (61.81 s)
********************************************************************************************************************************
[root@hp-dl385g10-14 kar]# python3 ConfigTest.py --testcase=migrate.default.tcp.default,migrate.default.tcp.with_filter_off.with_post_copy,migrate.default.mig_cancel,migrate.default.unix.default,migrate.default.exec.gzip_exec --guestname=RHEL.8.3.0 --clone=no --customsparam='qemu_force_use_static_incoming_expression=yes'
 (1/5) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.migrate.tcp: PASS (155.37 s)
 (2/5) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.migrate.tcp.with_filter_off.with_post_copy: PASS (61.80 s)
 (3/5) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.migrate.unix: PASS (152.24 s)
 (4/5) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.migrate.exec.gzip_exec: PASS (199.88 s)
 (5/5) Host_RHEL.m8.u3.product_av.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.3.0.x86_64.io-github-autotest-qemu.migrate.mig_cancel: PASS (57.88 s)
```

**Comment by ldoktor** [comment]
I can see the issue commented by @luckyh fixed and 2 acks, let me merge this

**Comment by luckyh** [comment]
> I can see the issue commented by @luckyh fixed and 2 acks, let me merge this

@ldoktor thanks.

**Comment by ldoktor** [comment]
> > I can see the issue commented by @luckyh fixed and 2 acks, let me merge this
> 
> @ldoktor thanks.

Big thanks to @YongxueHong as I was looking forward to this patch for a long time!

**Comment by YongxueHong** [comment]
> > > I can see the issue commented by @luckyh fixed and 2 acks, let me merge this
> > 
> > 
> > @ldoktor thanks.
> 
> Big thanks to @YongxueHong as I was looking forward to this patch for a long time!

Hi @ldoktor 
You are welcome.

**Pull request merged**